### PR TITLE
Merge ConnectionSpecification + Role -> Role

### PR DESCRIPTION
**Pull request by eileencodes**
In order to move schema_cache off of DatabaseConfiguration we needed to
make role accessible on pool.

While looking at `ConnectionSpecification`, `Role`, and `DatabaseConfig` John
and I noticed that this could be achieved by merging `ConnectionSpecification`
and `Role` into one `Role` class. This allows us to eliminate the `spec`
concept which is confusing. `spec` is a private method so renaming
to `resolve_role` is ok.

In the `Role` class we took `name` (renamed to `connection_specification_name`
for clarity since it's not a `role` name) and `db_config` from
`ConnectionSpecification` and the `pool` methods from `Role` and combined
them into one `Role` class.

This feels a lot cleaner to us because it clarifies the purposes of the
classes/methods/variables, and makes it easier to drop
`connection_specification_name` keyed on the parent class in the future.

There are a lot of changes in here but the majority of them are find and
replace `spec` -> `role`, `spec.name` ->
`role.connection_specification_name`, `Resolver#spec` ->
`Resolver#resolve_role`.

This PR also moves the `schema_cache` from `DatabaseConfig` to the new
combined `Role` class.
    
Co-authored-by: John Crepezzi <john.crepezzi@gmail.com>

cc/ @seejohnrun @rafaelfranca @tenderlove @jhawthorn @casperisfine @etiennebarrie

### CODE REVIEW

**Comment by eileencodes** [comment]
I don't know if this is where you all at Shopify were planning on going with the code around ConnSpec and Role, but it started to feel like we had 2 classes doing _kind of_ the same job. Happy to discuss more but wanted to open this PR as a discussion point and to see how the code feels when merging the 2 classes.

**Comment by rafaelfranca** [comment]
@Edouard-chin could you also review this to see if this would break the PR you are working on.

But I'm very happy with this direction.

**Review by rafaelfranca — APPROVED** [review]

**Comment by byroot** [comment]
On mobile so only did a superficial review, but yes it looks a lot like where I was trying to get at. If there is no urgency I’ll do a full review Sunday/Monday

**Comment by Edouard-chin** [comment]
ACK, I'll look if that doesn't conflict too much with the connection handling refactor we are doing with Jean. Will review later tonight or tomorrow !

**Review by Edouard-chin — APPROVED** [review]
Good idea ! 

> see if this would break the PR you are working on.

Nop it shouldn't conflict with what we are doing 👌

File: activerecord/lib/active_record/connection_adapters/role.rb [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+module ActiveRecord
```
Comment (casperisfine) [inline activerecord/lib/active_record/connection_adapters/role.rb]: 👍 to moving it to its own file. `connection_pool.rb` is huge and painful to work with.

File: activerecord/lib/active_record/connection_adapters/role.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+      class << self
+        def discard_pools!
+          INSTANCES.each_key(&:discard_pool!)
+        end
+      end
+
+      def initialize(connection_specification_name, db_config)
+        super()
+        @connection_specification_name = connection_specification_name
```
Comment (casperisfine) [inline activerecord/lib/active_record/connection_adapters/role.rb]: I guess this is ok for now, but I'd really like to get rid of `connection_specification_name`.

IMHO it's backwards. It's not the `Role / DbConfig` that should have access to the root model identifier (`AnimalsBase`), but the root model that should know what `Role / DbConfig` it connects to (`animals_replica`)

File: activerecord/lib/active_record/connection_adapters/connection_specification.rb [hunk 3]
```diff
@@ -1,15 +0,0 @@
-# frozen_string_literal: true
-
-require "uri"
-
-module ActiveRecord
-  module ConnectionAdapters
-    class ConnectionSpecification # :nodoc:
```
Comment (casperisfine) [inline activerecord/lib/active_record/connection_adapters/connection_specification.rb]: 🔥

**Review by casperisfine — APPROVED** [review]
💯

File: activerecord/lib/active_record/connection_adapters/role.rb (hunk 2 above)
Comment (eileencodes) [reply activerecord/lib/active_record/connection_adapters/role.rb]: Definitely want to get rid of this but for now this seemed like a good compromise. Continue using it but name it something obvious that screams "get me out of here" 😅

**Pull request merged**